pkg/relayer/miner/miner.go

```diff
@@ -0,0 +1,395 @@
+	_ = failedSubmitProofSessions
+	logging.LogErrors(ctx, filter.EitherError(ctx, eitherProvenSessions))
+}
+
+// validateConfigAndSetDefaults ensures that the miner has been configured with
+// a hasher and uses the default hasher if not.
+func (mnr *miner) validateConfigAndSetDefaults() error {
+	if mnr.hasher == nil {
+		mnr.hasher = defaultHasher
+	}
+	return nil
+}
+
+// mapMineRelay is intended to be used as a MapFn. It hashes the relay and compares
+// its difficulty to the minimum threshold. If the relay difficulty is sifficient,
+// it returns an either populated with the minedRelay value. Otherwise, it skips
+// the relay. If it encounters an error, it returns an either populated with the
+// error.
+func (mnr *miner) mapMineRelay(
+	_ context.Context,
+	relay *servicetypes.Relay,
+) (_ either.Either[*minedRelay], skip bool) {
+	relayBz, err := relay.Marshal()
+	if err != nil {
+		return either.Error[*minedRelay](err), true
+	}
+
+	// Is it correct that we need to hash the key while smst.Update() could do it
+	// since smst has a reference to the hasher
+	mnr.hasher.Write(relayBz)
+	relayHash := mnr.hasher.Sum(nil)
+	mnr.hasher.Reset()
+
+	if !protocol.BytesDifficultyGreaterThan(relayHash, defaultRelayDifficulty) {
+		return either.Success[*minedRelay](nil), true
+	}
+
+	return either.Success(&minedRelay{
+		Relay: *relay,
+		bytes: relayBz,
+		hash:  relayHash,
+	}), false
+}
+
+// mapAddRelayToSessionTree is intended to be used as a MapFn. It adds the relay
+// to the session tree. If it encounters an error, it returns the error. Otherwise,
+// it skips output (only outputs errors).
+func (mnr *miner) mapAddRelayToSessionTree(
+	_ context.Context,
+	eitherRelay either.Either[*minedRelay],
+) (_ error, skip bool) {
+	// Propagate any upstream errors.
+	relay, err := eitherRelay.ValueOrError()
+	if err != nil {
+		return err, false
+	}
+
+	// ensure the session tree exists for this relay
+	sessionHeader := relay.GetReq().GetMeta().GetSessionHeader()
+	smst, err := mnr.sessionManager.EnsureSessionTree(sessionHeader)
+	if err != nil {
+		log.Printf("failed to ensure session tree: %s\n", err)
+		return err, false
+	}
+
+	if err := smst.Update(relay.hash, relay.bytes, 1); err != nil {
+		log.Printf("failed to update smt: %s\n", err)
+		return err, false
+	}
+
+	// Skip because this map function only outputs errors.
+	return nil, true
+}
+
+// mapWaitForOpenClaimWindow is intended to be used as a MapFn. It calculates and
+// waits for the earliest block height at which it is safe to claim and returns
+// the session when it should be claimed.
+func (mnr *miner) mapWaitForOpenClaimWindow(
+	ctx context.Context,
+	session relayer.SessionTree,
+) (_ relayer.SessionTree, skip bool) {
+	mnr.waitForEarliestCreateClaimHeight(
+		ctx, session.GetSessionHeader().GetSessionEndBlockHeight(),
+	)
+	return session, false
+}
+
+// waitForEarliestCreateClaimHeight returns the earliest block height at which a
```

Comment:
The comment says `returns the earliest block height` but this function doesn't return anything, so it's a bit confusing & misleading. Can you clarify what you intended?